src/shogun/util/factory.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Heiko Strathmann
+ */
+#ifndef FACTORY_H_
+#define FACTORY_H_
+
+#include <shogun/base/class_list.h>
+#include <shogun/features/DenseFeatures.h>
+#include <shogun/io/CSVFile.h>
+#include <shogun/io/SGIO.h>
+#include <shogun/labels/DenseLabels.h>
+
+namespace shogun
+{
+
+	CKernel* kernel(const std::string& name);
+	CMachine* machine(const std::string& name);
+	CFeatures*
+	features(CFile* file, EPrimitiveType primitive_type = PT_FLOAT64);
+	template <class T>
+	CFeatures* features(SGMatrix<T> mat);
+	CLabels* labels(CFile* file);
+
+	CFile* csv_file(std::string fname, char rw = 'r');
+
+#define BASE_CLASS_FACTORY(T, factory_name)                                    \
+	T* factory_name(const std::string& name)                                   \
+	{                                                                          \
+		return create_object<T>(name.c_str());                                 \
+	}
+
+	BASE_CLASS_FACTORY(CKernel, kernel)
+	BASE_CLASS_FACTORY(CMachine, machine)
+
+	template <class T>
+	CFeatures* features(SGMatrix<T> mat)
+	{
+		CFeatures* features = new CDenseFeatures<T>(mat);
+		SG_REF(features);
+		return features;
+	}
+
+	CFeatures* features(CFile* file, EPrimitiveType primitive_type)
+	{
+		REQUIRE(file, "No file provided.\n");
+		CFeatures* result = nullptr;
+
+		if (dynamic_cast<CCSVFile*>(file))
+		{
+			switch (primitive_type)
+			{
+			case PT_FLOAT64:
+				result = new CDenseFeatures<float64_t>();
+				break;
+			default:
+				SG_SNOTIMPLEMENTED
+			}
+			result->load(file);
+		}
+		else
+			SG_SERROR("Cannot load features from %s.\n", file->get_name());
+
+		SG_REF(result);
+		return result;
+	}
+
+	CLabels* labels(CFile* file)
```

Comment:
@vigsterkr @lisitsyn this is the stuff behind the factory API

Reply:
I like the approach. It would become the main interface.